src/experimental/collection_new_iterable_model/shared/iterator/impl/RedBlackTreeIterator.hpp

```diff
@@ -0,0 +1,699 @@
+        AbstractTreeIterator < __ElementType > :: AbstractTreeIterator (                        // NOLINT(bugprone-reserved-identifier)
+                AbstractTreeIterator const & iterator
+        ) noexcept :
+                _pRoot ( iterator._pRoot ),
+                _pCurrentNode ( iterator._pCurrentNode ) {
+
+        }
+
+
+        template < typename __ElementType >                                // NOLINT(bugprone-reserved-identifier)
+        AbstractTreeIterator < __ElementType > :: AbstractTreeIterator (    // NOLINT(bugprone-reserved-identifier)
+                AbstractTreeIterator && iterator )
+        noexcept :
+            _pRoot        ( cds :: exchange ( iterator._pRoot, nullptr ) ),
+            _pCurrentNode ( cds :: exchange ( iterator._pCurrentNode, nullptr ) ) {
+
+        }
+
+
+        template < typename __ElementType >                             // NOLINT(bugprone-reserved-identifier)
+        auto AbstractTreeIterator < __ElementType > :: operator * (     // NOLINT(bugprone-reserved-identifier)
+        ) const noexcept -> __ElementType & {
+
+            return this->_pCurrentNode->_data;
+        }
+
+
+        template < typename __ElementType >                             // NOLINT(bugprone-reserved-identifier)
+        auto AbstractTreeIterator < __ElementType > :: operator -> (    // NOLINT(bugprone-reserved-identifier)
+        ) const noexcept -> __ElementType * {
+
+            return & this->_pCurrentNode->_data;
+        }
+
+
+        template < typename __ElementType >                             // NOLINT(bugprone-reserved-identifier)
+        auto AbstractTreeIterator < __ElementType > :: operator == (    // NOLINT(bugprone-reserved-identifier)
+                AbstractTreeIterator const & iterator
+        ) const noexcept -> bool {
+
+            return this->_pCurrentNode == iterator._pCurrentNode;
+        }
+
+
+        template < typename __ElementType >                             // NOLINT(bugprone-reserved-identifier)
+        auto AbstractTreeIterator < __ElementType > :: operator != (    // NOLINT(bugprone-reserved-identifier)
+                AbstractTreeIterator const & iterator
+        ) const noexcept -> bool {
+
+            return ! this->operator== ( iterator );
+        }
+
+
+        template < typename __ElementType >                                             // NOLINT(bugprone-reserved-identifier)
+        AbstractTreeIterator < __ElementType > :: operator bool () const noexcept {     // NOLINT(bugprone-reserved-identifier)
+
+            return this->_pCurrentNode != nullptr;
+        }
+
+
+        template < typename __ElementType >                                                                 // NOLINT(bugprone-reserved-identifier)
+        auto AbstractTreeIterator < __ElementType > :: successor () noexcept -> AbstractTreeIterator & {    // NOLINT(bugprone-reserved-identifier
+
```

Comment:
extract endNode as local const 

```cpp
auto const endNode = __hidden :: __impl :: __endNode < __ElementType > ();
if ( ... != endNode ) {
    ...
}
```

After investigation into usage in ++ and --
rename these two functions. Successor/Predecessor suggests returnal of  said value, immutable to caller object. Suggestion: advance / reverse or moveForward / moveBackward